README
blx::xsdsql
==========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

	XML::Parser 	=> '1.21' 
	XML::Writer 	=> '0.600'
	DBI	    		=> '1.58'
    Test::Database	=> '1.11.1'
    
    
For the connection to a database  the module DBI (version 1.58 or better) 
and the driver of the appropriate database are  required.
The database supported is only oracle and postgresql

After the installation the directory $PREFIX/bin contain two example of programs:
	xsd2sql - convert a xsd file into a database schema (text file with prefix .sql)
	xml.pl  - load | write | unload  a xml file from / to the database  


COPYRIGHT AND LICENCE
 
Copyright (C) 2010 by  lorenzo bellotti

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension blx::xsdsql

0.01 July  20 18:24 2010
	- original version; created by h2xs 1.23 with options
		-XA -b 5.8.0 -n blx::xsdsql::ut

0.02 December 09  18:48 2010
	- add TABLE_PREFIX param to blx::xsdsql::parser::parsefile
	- exclude dirs if name begin with a point into blx::xsdsql::parser::parsefile::get_db_namespaces
	- exclude dirs if name begin with a point into blx::xsdsql::parser::generator::get_namespaces
	- new commands display_namespaces,drop_sequence,create_sequence in xsd2sql.pl
	- correct the crash if the command is unknow in xsd2sql.pl
	- new options t,l,p,w in xsd2sql.pl
	- generic internal params restyle ( Es: my ($self,%params)=@_)	 in subs
	- add LEVEL_FILTER and TABLES_FILTER  params to blx::xsdsql::generator
	- add params SCHEMA_NAME and SCHEMA_INSTANCE to blx::xsdsql::xml::new and blx::xsdsql::xml::write
	- implement a new test suite
	- implement a xsd groups
	
0.03 January 04  19:04 2011  
	- correct bugs in  documentation
	- new commands drop_view, create_view 
	- implement drop view and create view classes 
	- change email address of the author 
	- minor bugs corretted  

0.04 January 20 20:08 2011
	- correct bugs in documentation
	- minor bugs corretted 
	- new commands drop_dictionary,create_dictionary,insert_dictionary in xsd2sql.pl
	- correct bug in create_view for oracle
	- make clean delete temporary files in test directory
    
0.05 March 31 17:07 2011 
	- complete restyle of parsing algorithms
	- minor bugs corretted 
	
0.06 June 11 12:00 2011
	- xsd group implemented 
	- a first implementation of namespace 
	- new options in test.pl/xsd2sql/xml.pl
	- bugs corrected 
	 
	
0.06.1 June 16 16:13 2011
	- corrected algo bugs on param ROOT_TAG_PARAMS of  method blx::xsdsql::xml::write

0.06.2 -  July 20 18:12 2011
	- corrected max int on parser 
	- bug in write mandatory note with value null string
	- create_view: in the header of views converted comment PATH to ROOTPATH
	- bug on choice - a default minOccurs for a choice node is 0 and not 1
	- bug on choice into a node with maxOccurs > 1
	- change the param CHOISE into CHOICE  in all modules
	- correct documentation
	- add this attribute in class blx::xsdsql::xml::generic::table and new columns in table_dictionary:
		PARENT_PATH,IS_ROOT_TABLE,IS_UNPATH,IS_INTERNAL_REF,VIEW_NAME
	- new method blx::xsdsql::xml::generic::table::reset_columns
	- revised the method  blx::xsdsql::xml::generic::table::add_columns
	- new param FORCE in blx::xsdsql::xml::generic::column::get_sql_name
	- new attribute CHOICE for class blx::xsdsql::xml::generic::column and new column in column_dictionary
	- flat the columns of a group with maxoccurs <= 1 into the parent table
	- new option -X in test.pl (modality change, test.pl use for default xmldiff if is installed)
	- new option -S in test.pl (modality change, for default test.pl esecute xsd2sql in one step) 
	- new param NO_FLAT_GROUPS in blx::xsdsql::parser::parsefile
	- correct bug on join a table with maxoccurs > 1 in create view
	- parser restyling code

0.07.0	- September 5 19:15 2011
	- simpleContent implementation
	- column  restyling code
	- table restyling code
	- new attribute ATTRIBUTE for class blx::xsdsql::xml::generic::column and new column in column_dictionary
	- new method   blx::xsdsql::path_map::resolve_attributes
	- attribute implementation
	- correct documentation
	- restyle structure  of table dictionary 
	- obsolete params TABLE_IS_TYPE,SIMPLE_TYPE,COMPLEX_TYPE,SIMPLE_CONTENT,GROUP_TYPE in blx::xsdsql::xml::generic::table
	- new param XSD_TYPE in  blx::xsdsql::xml::generic::table
	- correct bug on xseq choice

0.07.1	- September 7 18:05 2011
	- xsd2sql.pl -	correct warning parser command if $ARGV[0] is not set
					add attribute EMIT_SCHEMA_DUMPER 
	- exclusion system attribute for nodes at level 1
	- correct some tests because  not passed if is use diff command  

0.08 - <no_date> 
    - internal version - not published

0.09 - March 6 18:05 2012
	- new algorithmics for schema parsing (RINCHI::Xmlschema removing)
	- multiplies namespace support
	- support for import,fractionDigit,attributes,custom_header,element_ref,attribute_ref
    - better algorithm for generate names

0.09.1 - May 10 18:05 2012 
    - new syntax for perl 5.14

0.09.2 - November 27 - 17:19 2012
    - package in form tar.gz for win users (suggest from MITHALDU@cpan.org)


0.09.3 - January 20 - 18:00 2013 
	- new class method blx::xsdsql::dbconn::get_info
	- the method  blx::xsdsql::dbconn::get_application_avaliables is deprecated, use the new method get_application_availables
	- the class methods get_applications_classname,get_database_availables,get_application_avaliables 
	  of the class blx::xsdsql::dbconn are cached and return a sorted array 
	- the regression test use Test::Database for default (suggest from  CHORNY@cpan.org)
	- test.pl - new option D (do not use Test::Database for connection but the old method)
	- test.pl - new option F (do not clean the database after tests end)
	- isql.pl - new option e (use Test::Database for connection)
	- xml.pl  - new option e (use Test::Database for connection)
	- correct bug to retrieve connection string in blx::xsdsql::dbconn::oracle::dbi
    - adding META.yml (suggest from  CHORNY@cpan.org)
    - for tests xmldiff is mandatory


Makefile.PL
#!/usr/bin/env perl
use 5.008000;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use Getopt::Long qw(GetOptions);
use File::Copy;

GetOptions(\my %opt,
   'no-programs|n'
) or do {
    die "Usage: $0 [--no-programs|-n] \n";
};

my @prog;
push(@prog, qw(xml.pl xsd2sql.pl ))
    unless $opt{'no-programs'} || grep /^LIB=/, @ARGV;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

#print STDERR  "---->",join(" ", map "bin/$_", grep /^[A-Z]+$/, @prog),"\n";

WriteMakefile(
	NAME              => 'blx::xsdsql'
	,VERSION_FROM      => 'lib/blx/xsdsql/ut.pm'
	,PREREQ_PM         => { 
						  'XML::Parser' 	=> '1.21' 
						, 'XML::Writer' 	=> '0.600'
						, 'DBI'			=> '1.58'
						, 'Test::Database' 	=> '1.11'
	}
	,LICENSE => "perl"
	,EXE_FILES => [ map "bin/$_", @prog ]
	,($] >= 5.005 ?     ## Add these new keywords supported since 5.005
		(
			ABSTRACT 	=> 'xml schema (xsd) converter to relational database  objects (database supported are oracle and postgresql); xml put and get from/to the database'
			,AUTHOR         => 'lorenzo bellotti <pauseblx@gmail.com>'
		) : ()
	)
	,clean => { FILES => "  *.old t/xml_*/*.sql   t/xml_*/*.diff t/xml_*/*.tmp  t/xml_*/.step* t/.strconn " }
);


move("MYMETA.yml","META.yml");
exit 0;

